library/time-library/src/main/java/uk/gov/gchq/gaffer/time/BoundedTimestampSet.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2017 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.gaffer.time;
+
+import com.yahoo.sketches.sampling.ReservoirLongsUnion;
+import org.apache.commons.lang3.StringUtils;
+import uk.gov.gchq.gaffer.commonutil.CommonTimeUtil;
+import uk.gov.gchq.gaffer.commonutil.CommonTimeUtil.TimeBucket;
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.OptionalLong;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+/**
+ * A <code>BoundedTimestampSet</code> is an implementation of {@link TimestampSet} that can contain a maximum number
+ * N of timestamps. If more than N timestamps are added then a uniform random sample of size approximately N of the
+ * timestamps is retained.
+ *
+ * <p>This is useful in avoiding the in-memory or serialised size of the set of timestamps becoming too
+ * large. If less than N timestamps are added then the timestamps are stored in a {@link RBMBackedTimestampSet}. If
+ * more than N timestamps are added then a uniform random sample of size N of the timestamps is retained.
+ */
+public class BoundedTimestampSet implements TimestampSet {
```

Comment:
This class is not json serialisable. To enable it to be used over the rest api, it needs to be json serialisable, i.e add a default constructor, getters and setters.

Perhaps add a simple test like this:

```java
    @Test
    public void shouldSerialiseAndDeserialiseOperation() throws SerialisationException {
        // Given
        final JSONSerialiser serialiser = new JSONSerialiser();
        final BoundedTimestampSet boundedTimestampSet = new BoundedTimestampSet(CommonTimeUtil.TimeBucket.SECOND, 10);
        IntStream.range(0, 20)
                .forEach(i -> {
                    boundedTimestampSet.add(Instant.ofEpochMilli(i * 1000L));
                    if (i <= 9) {
                        assertEquals(BoundedTimestampSet.State.NOT_FULL, boundedTimestampSet.getState());
                    } else {
                        assertEquals(BoundedTimestampSet.State.SAMPLE, boundedTimestampSet.getState());
                    }
                });


        // When
        byte[] json = serialiser.serialise(boundedTimestampSet, true);
        final BoundedTimestampSet deserialisedObj = serialiser.deserialise(json, BoundedTimestampSet.class);

        // Then
        assertEquals(boundedTimestampSet, deserialisedObj);
    }
```